### PR TITLE
Fix the appearance of the line number and highlight for the first line when there is no text in the entire text editor.

### PR DESCRIPTION
**Pull request by HassanTaleb90**
I think we should reconsider the approach `firstBaseline: locationForFirstCharacter.y + baselineYOffset`. Perhaps using `firstBaseline: (typingLineHeight + characterHeight) / 2` would be better (see image below). 

For this example, I used a very large font size of 288 points.

Screenshot 2024-08-22 at 6 09 30 AM

Additionally, it might be better to increase the line number font size to match the text size, similar to how Xcode does it.

https://github.com/user-attachments/assets/a4ff6e4d-2102-437e-93ff-7441ad50b9c3

### CODE REVIEW

File: Sources/STTextViewAppKit/STTextView+Gutter.swift [hunk 1]
```diff
@@ -72,7 +72,49 @@ extension STTextView {
         let selectedLineTextAttributes: [NSAttributedString.Key: Any] = [
             .foregroundColor: (gutterView.selectedLineTextColor ?? gutterView.textColor).cgColor
         ]
-
+        
+        if viewportRange.isEmpty {
```
Comment (krzyzanowskim) [inline Sources/STTextViewAppKit/STTextView+Gutter.swift:76]:
```suggestion
        if textLayoutManager.documentRange.isEmpty {
```

File: Sources/STTextViewAppKit/STTextView+Gutter.swift [hunk 2]
```diff
@@ -72,7 +72,49 @@ extension STTextView {
...
+            let lineNumber = 1
+            
+            var lineTextAttributes: [NSAttributedString.Key: Any] = [
+                .font: gutterView.font,
+                .foregroundColor: NSColor.secondaryLabelColor
+            ]
```
Comment (krzyzanowskim) [inline Sources/STTextViewAppKit/STTextView+Gutter.swift:82]:
```suggestion
            var effectiveLineTextAttributes = lineTextAttributes
```
`lineTextAttributes` is already defined above.

File: Sources/STTextViewAppKit/STTextView+Gutter.swift [hunk 3]
```diff
@@ -72,7 +72,49 @@ extension STTextView {
...
+            
+            let isLineSelected = gutterView.highlightSelectedLine && !selectedLineTextAttributes.isEmpty
+            if isLineSelected {
+                lineTextAttributes.merge(selectedLineTextAttributes, uniquingKeysWith: { _, new in new })
```
Comment (krzyzanowskim) [inline Sources/STTextViewAppKit/STTextView+Gutter.swift:86]:
```suggestion
                effectiveLineTextAttributes.merge(selectedLineTextAttributes, uniquingKeysWith: { _, new in new })
```

File: Sources/STTextViewAppKit/STTextView+Gutter.swift [hunk 4]
```diff
@@ -72,7 +72,49 @@ extension STTextView {
...
+            }
+    
+            let characterHeight = NSAttributedString(string: "\(lineNumber)", attributes: lineTextAttributes)
```
Comment (krzyzanowskim) [inline Sources/STTextViewAppKit/STTextView+Gutter.swift:89]:
```suggestion
            let characterHeight = NSAttributedString(string: "\(lineNumber)", attributes: effectiveLineTextAttributes)
```

File: Sources/STTextViewAppKit/STTextView+Gutter.swift [hunk 5]
```diff
@@ -72,7 +72,49 @@ extension STTextView {
...
+                .boundingRect(
+                    with: CGSize(width: CGFloat.greatestFiniteMagnitude, height: CGFloat.greatestFiniteMagnitude),
+                    options: [.usesDeviceMetrics],
+                    context: nil
+                ).height
+
```
Comment (krzyzanowskim) [inline Sources/STTextViewAppKit/STTextView+Gutter.swift:95]: check out `updateSelectedLineHighlight()` how to get the same metric here

**Review by krzyzanowskim — CHANGES_REQUESTED** [review]
thank you! it looks good overall. I think we should handle metric in a different way (and consistent with other places)

**Comment by HassanTaleb90** [comment]
In Xcode, I noticed that there is always a line break at the end of the file. So, I'm not sure if it would be better to follow the same approach. What do you think?

Screenshot 2024-08-22 at 7 47 55 PM

Screenshot 2024-08-22 at 7 59 18 PM

**Review by HassanTaleb90 — COMMENTED** [review]
👍

**Comment by krzyzanowskim** [comment]
> In Xcode, I noticed that there is always a line break at the end of the file

this last line is an `extraLineFragment` in our nomenclature. When it comes to the whitespace, better keep it in sync with the characters. if there's none, we should not add.

File: Sources/STTextViewAppKit/STTextView+Gutter.swift (hunk 5 above)
Comment (HassanTaleb90) [reply Sources/STTextViewAppKit/STTextView+Gutter.swift:95]: Maybe like this, but in our case the viewportRange is empty
` var characterHeight: CGFloat = 0
            textLayoutManager.enumerateTextLayoutFragments(in: viewportRange) { layoutFragment in
                for lineFragment in layoutFragment.textLineFragments {
                    let lineFragmentHeight = lineFragment.typographicBounds.height
                    characterHeight = lineFragmentHeight
                    return false
                }
                return true
            }
         `

**Comment by krzyzanowskim** [comment]
done in 141b60b313a4e58668971bbecb3165ca27ec200f

**Pull request closed without merging**